mcs/class/corlib/System.Runtime.InteropServices/Marshal.cs

```diff
@@ -1897,5 +1897,84 @@ internal static void SetLastWin32Error (int error)
 			throw new NotSupportedException();
 		}
 #endif
+
+		internal struct MarshalerInstanceKey {
+			public Type Type;
+			public string Cookie;
+
+			public override int GetHashCode () {
+				int result = 0;
+				if (Type != null)
+					result = Type.GetHashCode();
+				if (Cookie != null)
+					result ^= Cookie.GetHashCode();
+				return result;
+			}
+		}
+
+		internal class MarshalerInstanceKeyComparer : IEqualityComparer<MarshalerInstanceKey> {
+			public bool Equals (MarshalerInstanceKey lhs, MarshalerInstanceKey rhs) {
+				return (lhs.Type == rhs.Type) && (lhs.Cookie == rhs.Cookie);
+			}
+
+			public int GetHashCode (MarshalerInstanceKey key) {
+				return key.GetHashCode ();
+			}
+		}
+
+		internal static Dictionary<MarshalerInstanceKey, ICustomMarshaler> MarshalerInstanceCache = 
+			new Dictionary<MarshalerInstanceKey, ICustomMarshaler> (new MarshalerInstanceKeyComparer ());
+
+		internal static ICustomMarshaler GetCustomMarshalerInstance (Type type, string cookie) {
+			var key = new MarshalerInstanceKey {
+				Type = type,
+				Cookie = cookie
+			};
+
+			ICustomMarshaler result;
+			bool gotExistingInstance;
+			lock (MarshalerInstanceCache)
+				gotExistingInstance = MarshalerInstanceCache.TryGetValue (key, out result);
+
+			if (!gotExistingInstance) {
+				MonoMethod getInstanceMethod;
+				try {
+					getInstanceMethod = (MonoMethod)type.GetMethod (
+						"GetInstance", BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.InvokeMethod,
+						null, new Type[] { typeof(string) }, null
+					);
+				} catch (AmbiguousMatchException) {
+					throw new ApplicationException ($"Custom marshaler '{type.FullName}' implements multiple static GetInstance methods that take a single string parameter.");
```

Comment:
ApplicationException is not right exception for BCL code

Reply:
It's what this code throws in other implementations

Reply:
oh, really. that's strange